src/realm/object-store/sync/app.cpp

```diff
@@ -384,23 +390,54 @@ void App::configure_route(const std::string& host_url, const std::optional<std::
     if (ws_host_url && ws_host_url->length() > 0) {
         m_ws_host_url = *ws_host_url;
     }
-    // Otherwise, convert the host url to a websocket host url (http[s]:// -> ws[s]://)
+    // Otherwise, convert the host url to a websocket host url
     else {
-        m_ws_host_url = m_host_url;
-        if (m_ws_host_url.find("http") == 0) {
-            m_ws_host_url.replace(0, 4, "ws");
-        }
+        m_ws_host_url = App::create_ws_host_url(m_host_url);
     }
 
     // host_url is the url to the server: e.g., https://realm.mongodb.com or https://localhost:9090
-    // base_route is the baseline client api path: e.g. https://realm.mongodb.com/api/client/v2.0
+    // base_route is the baseline client api path: e.g. <host_url>/api/client/v2.0
     m_base_route = util::format("%1%2", m_host_url, s_base_path);
-    // app_route is the cloud app URL: https://realm.mongodb.com/api/client/v2.0/app/<app_id>
+    // app_route is the cloud app URL: <host_url>/api/client/v2.0/app/<app_id>
     m_app_route = util::format("%1%2/%3", m_base_route, s_app_path, m_config.app_id);
-    // auth_route is cloud app auth URL: https://realm.mongodb.com/api/client/v2.0/app/<app_id>/auth
+    // auth_route is cloud app auth URL: <host_url>/api/client/v2.0/app/<app_id>/auth
     m_auth_route = util::format("%1%2", m_app_route, s_auth_path);
 }
 
+// Create a temporary websocket URL domain using the given host URL
+// This updates the URL based on the following assumptions:
+// If the URL doesn't start with 'http' => <host-url>
+// http[s]://[region-prefix]realm.mongodb.com => ws[s]://ws.[region-prefix]realm.mongodb.com
+// http[s]://[region-prefix]services.cloud.mongodb.com => ws[s]://[region-prefix].ws.services.cloud.mongodb.com
+// All others => http[s]://<host-url> => ws[s]://<host-url>
+std::string App::create_ws_host_url(const std::string_view host_url)
+{
+    constexpr static std::string_view s_orig_base_domain = "realm.mongodb.com";
+    constexpr static std::string_view s_new_base_domain = "services.cloud.mongodb.com";
+
+    // Doesn't start with http, just return provided string
+    if (host_url.substr(0, 4) != "http") {
+        return std::string(host_url);
+    }
+    // If it starts with 'https' then ws url will start with wss
+    bool https = host_url[4] == 's';
+    size_t prefix_len = std::char_traits<char>::length("http://") + (https ? 1 : 0);
+    std::string_view prefix = https ? "wss://" : "ws://";
+
+    // http[s]://[region-prefix]realm.mongodb.com => ws[s]://ws.[region-prefix]realm.mongodb.com
+    if (host_url.find(s_orig_base_domain) != std::string_view::npos) {
```

Comment:
should we check host_url ends with s_orig_base_domain? (same for s_new_base_domain)

Reply:
This function is designed to also work in case there is extra info (path, query vars, etc.) after the base URL value, so `host_url` may not necessarily end with the orig base domain.

Reply:
I see. I mainly added the comments due to the examples you provided.

Reply:
I updated the comment strings to reflect the optional path at the end